### ISSUE TITLE
UDP forward over SOCKS fails!

### ISSUE THREAD

**Issue by lars18th** [ginuerzh/gost#62]
Hi,

Here I describe my setup with "gost 2.3-dev" (last release):

Client: 192.168.1.33
GOST local: 192.168.1.35
GOST remote: 10.10.8.1
Server: 10.10.5.2

All machines are Linux Debian x64, and Client & GOST local are in the same subnet, and GOST remote & Server are in another subnet. Both GOST instances can connect between them through a router.

The I start the two instances of gost with:

`#  ./gost -L=socks://10.10.8.1:8701 -logtostderr -v=5`

`# ./gost -L=udp://192.168.1.35:1701/10.10.5.2:1701 -F=socks://10.10.8.1:8701 -logtostderr -v=5`

So, as you can imagine the idea is create a UDP forward from 192.168.1.35:1701 to 10.10.5.2:1701 using the tunnel.

Then I start a L2TP server in 10.10.5.2 listening at 1701, and the L2TP client in 192.168.1.33 using as the remote address "192.168.1.35:1701".

The problem is: The first packet from the client arrives to the server. The server responds, and the packet is forwarded to the client. Then the client sends more packets... and they arrive to the server, but **no more responses from the server are received by the client**.

I debug the problem using also "tcpdump -n udp port 1701 -vv" on both: the gost client and gost server. Here some logs:

**GOST SERVER LOG** (IP: 10.10.8.1):

```
I1213 19:59:39.115221   15796 socks.go:189] [udp] 192.168.1.35:60036 - 10.10.5.2:1701
I1213 19:59:39.115431   15796 socks.go:517] [udp] 192.168.1.35:60036 <-> 10.10.5.2:1701
I1213 19:59:54.478050   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 112
I1213 19:59:54.481132   15796 socks.go:551] [udp-tun] 192.168.1.35:60036 <<< 10.10.5.2:1701 length: 108
I1213 19:59:54.513514   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 20
I1213 19:59:55.514881   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 20
I1213 19:59:56.516526   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 20
I1213 19:59:58.519344   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 20
I1213 20:00:02.523149   15796 socks.go:541] [udp-tun] 192.168.1.35:60036 >>> 10.10.5.2:1701 length: 20
```

**GOST CLIENT LOG** (IP: 192.168.1.35):

```
I1213 19:59:39.133469     799 forward.go:192] [udp] 192.168.1.35:1701 <-> 10.10.5.2:1701
I1213 19:59:54.463943     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 112
I1213 19:59:54.498766     799 forward.go:225] [udp-tun] 192.168.1.35:1701 <<< 10.10.5.2:1701 length: 108
I1213 19:59:54.500110     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 20
I1213 19:59:55.501567     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 20
I1213 19:59:56.502597     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 20
I1213 19:59:58.504795     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 20
I1213 20:00:02.509068     799 forward.go:206] [udp-tun] 192.168.1.35:1701 >>> 10.10.5.2:1701 length: 20

```

With the TCPDUMP tool I see that the first response from the Server (first return packet) is comming to the client... I see the packet in the gost remote and in the gost local. But after that point, all packets from the server are also shown in the gost remote machine, but none of them appear in the gost local. However, all packets from the client are present in both machines. In the logs you can show the real traffic between client-server. I can't debug the incomming/outcomming UDP packets as "gost" can't handle this. However, with the tcpdump tool you can show the packets.

**So, the BUG IS THIS: After the first response from the UDP server, all other packets in return are discarded!**

I can confirm that using other UDP forwading tools (like "delegated", for example) in the same scenario all works. So, the problem is with "gost".

Please, can you review the code and fix this bug?
Thank you!

**Comment by lars18th** [comment]
Hi,

Regarding this issue I isolated the problem with the read of the UDP response from the server in the gost remote...

After some time running the tunnel, the GOST REMOTE logs some lines like this:

```
I1213 21:02:00.287919   15908 socks.go:546] [udp] 192.168.1.35:40482 <- 10.10.5.2:1701 : read udp 10.10.8.1:59595->10.10.5.2:1701: i/o timeout
I1213 21:02:04.291047   15908 socks.go:546] [udp] 192.168.1.35:40482 <- 10.10.5.2:1701 : read udp 10.10.8.1:38578->10.10.5.2:1701: i/o timeout

```

As you can see the problem is with the gost remote that can't read the response from the server.

If I execute one "netstat -an | grep 1701" during the execution I see several (a lot) of UDP sockets open from the gost remote to the server... But this is an error! The gost server should use the same UDP source port for all packets send to the server. Instead of this, it seems that it creates a new one for each new packet that arrives from the tunnel.

Please, can you review the code?
Thank you!

**Comment by ginuerzh** [comment]
At first, the UDP local forwarding feature was primarily used for DNS lookup service to avoid DNS pollution, just one roundtrip per request. Because udp is connectionless, and the relationship between client and server is many-to-one, so we cannot use a single source port to send the requests.

But I'll try to map each client to a fixed port on the remote side in the next release.

**Comment by lars18th** [comment]
Hi,

Thank you for your support!

Now, I understand the current behaviour. If you "only" use it for DNS resolution, then the one-trip is OK. However, I suggest to explain and document the current behaviour. The description in the README doesn't explain this limitation.

Regarding a possible fix, your idea seems good. However, as I know the current protocol implementation knows if the UDP tunnel is running or not. Then, I feel it's more easy to map each running tunnel to a fixed udp port. You agree?

Thank you for this good tool!

**Comment by lars18th** [comment]
Hi ginuerzh,

I'm new to Go, and I try to implement a simple workaround: write all UDP packages over the same SOURCE port, like this:

```
socks.go#532

		go func() {
			b := make([]byte, LargeBufferSize)

+			sourceUdpAddr, err := net.ResolveUDPAddr("udp4", "0.0.0.0:1701")
+
-			relay, err := net.DialUDP("udp", nil, raddr)
+			relay, err := net.DialUDP("udp", sourceUdpAddr, raddr)

```

However, as this seems to work at start (I see packets forwarding in both directions), after some initial packets the flow stops and I see in the logs:

```
[1225 20:19:11.253423  142448 socks.go:539] [udp] 192.168.1.35:55106 -> 10.10.5.2:1701 : dial udp 0.0.0.0:1701->10.10.5.2:1701: bind: Only one use in each direction (protocol/network address/port)
```

Please, can you help me to fix this?
Thank you!

**Comment by ginuerzh** [comment]
this code block is a goroutine (like a thread), so each request arrives, a new goroutine will be created and run in the background. If the last request has not been done, then the conflict will occur (due to the definition of socket pair).

**Comment by lars18th** [comment]
Hi ginuerzh,

Oh, yes! "go func()"... Then, any idea?
Perhaps "create" the UDP socket (net.DialUPD() call) before the goroutine?

Please, help me!
Thank you!

**Comment by ginuerzh** [comment]
I have fixed this issue in 2.3-rc1, you can re-test it, but there is still another problem: #65 , you can continue to use SOCKS5 proxy to avoid this problem.

**Comment by ginuerzh** [comment]
Finally, the UDP port forwarding based on HTTP2 (issue #65) also works fine, 2.3-rc2 is released, I recommend you upgrade to this version.
